### PR TITLE
Update various api serializers to expose total number a card moves to review feedback

### PR DESCRIPTION
**Pull request by Andy-Nkumane**
Related issues: [please specify]

## Description:

What are you up to? Fill us in :)

## I solemnly swear that:

- [x] My code follows the style guidelines of this project
- [x] I have merged the develop branch into my branch and fixed any merge conflicts
- [x] I have performed a self-review of my own code
- [x] I have commented my code in hard-to-understand areas
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: backend/curriculum_tracking/serializers.py [hunk 1]
```diff
@@ -354,8 +365,18 @@ class Meta:
             "open_pr_count",
             "oldest_open_pr_updated_time",
             "repo_url",
+            "total_number_of_negative_reviews",
```
Comment (sheenarbw) [inline backend/curriculum_tracking/serializers.py]: maybe a better name would be:

number_of_times_card_moved_to_review_feedback.

If this was really the total number of negative reviews then the call would be very different.

File: backend/curriculum_tracking/serializers.py [hunk 2]
```diff
@@ -326,6 +330,13 @@ def get_users_that_reviewed_open_prs(self, instance):
     def get_users_that_reviewed_open_prs_emails(self, instance):
         return [o.email for o in instance.get_users_that_reviewed_open_prs()]
 
+    def get_total_number_of_negative_reviews(self, instance):
+        total_number_of_negative_reviews = len(
+            LogEntry.objects.filter(effected_user=instance.assignees.first(),
```
Comment (sheenarbw) [inline backend/curriculum_tracking/serializers.py]: This query gets the total number of times any card belonging to the user was bounced back.

You need to filter by the actual card we are looking at. 

If I have 2 cards:

card_1 bounced 2 times and card_2 bounced 3 times then the total bounces for card_1 is 2 and not 5.

Comment (sheenarbw) [reply backend/curriculum_tracking/serializers.py]: please write a test

File: backend/curriculum_tracking/serializers.py [hunk 3]
```diff
@@ -932,3 +956,11 @@ def get_reviewer_user_emails(self, instance):
     def get_status(self, instance):
         if instance.agile_card:
             return instance.agile_card.status
+
+    def get_total_number_of_negative_reviews(self, instance):
+        total_number_of_negative_reviews = len(
+            LogEntry.objects.filter(effected_user=instance.recruit_users.first(),
```
Comment (sheenarbw) [inline backend/curriculum_tracking/serializers.py]: not DRY.  This function has the same problem as the others

**Review by sheenarbw — CHANGES_REQUESTED** [review]
The logic is incorrect. Also it is not DRY. 

Please test your work.  I also think you'll need to update the existing tests so that they pass. The api test mixin will end up failing a bunch of stuff if you add extra fields that retuyrn falsy data by default.  If that sentence doesn't make sense, please run the tests

File: backend/curriculum_tracking/serializers.py [hunk 4]
```diff
@@ -327,6 +331,16 @@ def get_users_that_reviewed_open_prs(self, instance):
     def get_users_that_reviewed_open_prs_emails(self, instance):
         return [o.email for o in instance.get_users_that_reviewed_open_prs()]
 
+    def get_number_of_times_card_moved_to_review_feedback(self, instance):
+        if instance.recruit_project:
```
Comment (Samantha-Hampton) [inline backend/curriculum_tracking/serializers.py]: maybe we aren't interested in topics for this but there are some topics that require review, so then if they did get a negative review, it wouldn't come up here. Maybe we just need to check to make sure we aren't filtering out data we may want?

**Review by Samantha-Hampton — CHANGES_REQUESTED** [review]
I think it mostly looks ok, but maybe good to write that extra test Sheena was suggesting that checks that the correct values are coming out, eg:

If I have 2 cards:

card_1 bounced 2 times and card_2 bounced 3 times then the total bounces for card_1 is 2 and not 5.

File: backend/curriculum_tracking/tests/test_api_views.py [hunk 5]
```diff
@@ -447,6 +457,113 @@ def test_set_project_link(self):
...
+        self.assertEqual(response.status_code, 200)
+        card1.refresh_from_db()
+
+        card_moved_to_review_feedback_id = EventType.objects.filter(name=creators.CARD_MOVED_TO_REVIEW_FEEDBACK).first().id
+        c = LogEntry.objects.filter(
+            event_type=card_moved_to_review_feedback_id, 
+            effected_user=card1.assignees.first(),
+            object_1_id=card1.content_item.id,
+        ).last()
+        c.timestamp = c.timestamp - timedelta(days=2)
```
Comment (Samantha-Hampton) [inline backend/curriculum_tracking/tests/test_api_views.py]: I'm getting an error when I run the tests:

```
ERROR: test_get_number_of_times_card_moved_to_review_feedback (curriculum_tracking.tests.test_api_views.AgileCardViewsetTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/sam/Documents/Umuzi/Products/Tilde/backend/curriculum_tracking/tests/test_api_views.py", line 500, in test_get_number_of_times_card_moved_to_review_feedback
    c = LogEntry.objects.filter(
AttributeError: 'NoneType' object has no attribute 'timestamp'

----------------------------------------------------------------------
Ran 70 tests in 116.148s

FAILED (errors=1)
```

So it's not registering a matching log entry object, so there is no timestamp to update

**Review by Samantha-Hampton — CHANGES_REQUESTED** [review]

File: backend/curriculum_tracking/tests/test_api_views.py [hunk 6]
```diff
@@ -447,6 +457,98 @@ def test_set_project_link(self):
         project.refresh_from_db()
         self.assertEqual(project.link_submission, link_2)
 
+    def test_get_number_of_times_card_moved_to_review_feedback(self):
+        recruit = UserFactory()
```
Comment (Samantha-Hampton) [inline backend/curriculum_tracking/tests/test_api_views.py]: extremely minor, but this recruit item does not appear to be used, so can be deleted

**Review by Samantha-Hampton — CHANGES_REQUESTED** [review]

**Review by Rokhuda — APPROVED** [review]

File: backend/curriculum_tracking/serializers.py [hunk 7]
```diff
@@ -355,8 +367,21 @@ class Meta:
             "open_pr_count",
             "oldest_open_pr_updated_time",
             "repo_url",
+            "number_of_times_card_moved_to_review_feedback",
         ]
 
+    number_of_times_card_moved_to_review_feedback = serializers.SerializerMethodField("get_number_of_times_card_moved_to_review_feedback")
+
+    def get_number_of_times_card_moved_to_review_feedback(self, instance):
```
Comment (sheenarbw) [inline backend/curriculum_tracking/serializers.py:375]: This really doesn't look DRY to me. Please avoid copy-pasting whole functions.

File: backend/curriculum_tracking/serializers.py [hunk 8]
```diff
@@ -355,8 +367,21 @@ class Meta:
...
+        if instance.recruit_project:
+            number_of_times_card_moved_to_review_feedback = len(
+                LogEntry.objects.filter(effected_user=instance.assignees.first(), 
+                object_1_id=instance.recruit_project.pk,
```
Comment (sheenarbw) [inline backend/curriculum_tracking/serializers.py:379]: This query should be more explicit. 

We can have a recruit project with the same id as a repo. We can have a recruit project with the same id as a review as well.

Please also add object_1's content type to the query. Explicit is better than implicit.

File: backend/curriculum_tracking/serializers.py [hunk 9]
```diff
@@ -928,3 +956,12 @@ def get_reviewer_user_emails(self, instance):
     def get_status(self, instance):
         if instance.agile_card:
             return instance.agile_card.status
+
+    def get_number_of_times_card_moved_to_review_feedback(self, instance):
+        number_of_times_card_moved_to_review_feedback = len(
+            LogEntry.objects.filter(effected_user=instance.recruit_users.first(),
```
Comment (sheenarbw) [inline backend/curriculum_tracking/serializers.py:965]: please avoid copy paste. DRY code is better.

You can add a function to the RecruitProject model. Then just call the function as needed.

There is no reason to introduce this technical debt now.

File: backend/curriculum_tracking/tests/test_activity_log_entry_creators.py [hunk 10]
```diff
@@ -122,6 +122,7 @@ def test_start_project(self):
                 project_submission_type=ContentItem.LINK, template_repo=None
             ),
         )
+        creators.EventType.objects.get_or_create(name=creators.CARD_MOVED_TO_REVIEW_FEEDBACK)
```
Comment (sheenarbw) [inline backend/curriculum_tracking/tests/test_activity_log_entry_creators.py]: These tests shouldn't need to be updated.

If you make your sql queries more efficient then these changes wont be required.

File: backend/curriculum_tracking/serializers.py [hunk 11]
```diff
@@ -928,3 +956,12 @@ def get_reviewer_user_emails(self, instance):
...
+            object_1_id=instance.pk, 
+            event_type=EventType.objects.get(name='CARD_MOVED_TO_REVIEW_FEEDBACK')),
```
Comment (sheenarbw) [inline backend/curriculum_tracking/serializers.py]: This query results in multiple SQL queries being fired off. We could mke this more efficient by doing something like:

```
filter(event_type__name=CARD_MOVED_TO_REVIEW_FEEDBACK)
```

File: backend/curriculum_tracking/tests/test_api_views.py [hunk 12]
```diff
@@ -53,6 +55,7 @@ def verbose_instance_factory(self):
         card = factories.AgileCardFactory(recruit_project=project)
         card.reviewers.add(core_factories.UserFactory())
         card.assignees.add(core_factories.UserFactory())
+        creators.EventType.objects.get_or_create(name=creators.CARD_MOVED_TO_REVIEW_FEEDBACK)
```
Comment (sheenarbw) [inline backend/curriculum_tracking/tests/test_api_views.py]: These changes shouldn't be needed

**Review by sheenarbw — CHANGES_REQUESTED** [review]
A few things can be made more efficient. And the code is still not DRY

**Pull request closed without merging**